File: src/buyer.ts
import AcpClient, {
    AcpContractClientV2,
    AcpJobPhases,
    AcpJob,
    AcpMemo,
    AcpAgentSort,
    AcpGraduationStatus,
    AcpOnlineStatus,
} from "@virtuals-protocol/acp-node";
import dotenv from "dotenv";
dotenv.config({ path: ".env.local" });

const BUYER_WALLET_PRIVATE_KEY = process.env.BUYER_WALLET_PRIVATE_KEY as `0x${string}`;
const BUYER_ENTITY_ID = parseInt(process.env.BUYER_ENTITY_ID!);
const BUYER_AGENT_WALLET_ADDRESS = process.env.BUYER_AGENT_WALLET_ADDRESS as `0x${string}`;

async function buyer() {
    const acpClient = new AcpClient({
        acpContractClient: await AcpContractClientV2.build(
            BUYER_WALLET_PRIVATE_KEY,
            BUYER_ENTITY_ID,
            BUYER_AGENT_WALLET_ADDRESS
        ),
        onNewTask: async (job: AcpJob, memoToSign?: AcpMemo) => {
            if (
                job.phase === AcpJobPhases.NEGOTIATION &&
                memoToSign?.nextPhase === AcpJobPhases.TRANSACTION
            ) {
                console.log(`Paying for job ${job.id}`);
                await job.payAndAcceptRequirement();
                console.log(`Job ${job.id} paid`);
            } else if (
                job.phase === AcpJobPhases.TRANSACTION &&
                memoToSign?.nextPhase === AcpJobPhases.REJECTED
            ) {
                console.log(`Signing job ${job.id} rejection memo, rejection reason: ${memoToSign?.content}`);
                await memoToSign?.sign(true, "Accepts job rejection")
                console.log(`Job ${job.id} rejection memo signed`);
            } else if (job.phase === AcpJobPhases.COMPLETED) {
                console.log(`Job ${job.id} completed, received deliverable:`, job.deliverable);
            } else if (job.phase === AcpJobPhases.REJECTED) {
                console.log(`Job ${job.id} rejected by seller`);
            }
        }
    });

    // Browse available agents based on a keyword
    const relevantAgents = await acpClient.browseAgents(
        "Chromia's EVAL",
        {
            sort_by: [AcpAgentSort.SUCCESSFUL_JOB_COUNT],
            top_k: 5,
            graduationStatus: AcpGraduationStatus.NOT_GRADUATED,
            onlineStatus: AcpOnlineStatus.ALL,
        }
    );

    console.log("Relevant agents:", relevantAgents.length);

    // // Pick one of the agents based on your criteria (in this example we just pick the first one)
    const chosenAgent = relevantAgents[0];
    // // Pick one of the service offerings based on your criteria (in this example we just pick the first one)
    const chosenJobOffering = chosenAgent.jobOfferings[0];
    // console.log("Chosen job offering:", chosenJobOffering);

    const jobId = await chosenJobOffering.initiateJob(
        {
            "coin_id": "chromaway",
            "details": "$CHR token, gas free on our chain!"
          },
        BUYER_AGENT_WALLET_ADDRESS, // evaluator address
        new Date(Date.now() + 1000 * 60 * 60 * 24) // expiredAt
    );

    console.log(`Job ${jobId} initiated`);
}

buyer();

File: acp/buyer.py
import logging
import threading
from datetime import datetime, timedelta
from typing import Optional

from dotenv import load_dotenv
from virtuals_acp.memo import ACPMemo
from virtuals_acp.client import VirtualsACP
from virtuals_acp.job import ACPJob
from virtuals_acp.models import (
    ACPAgentSort,
    ACPJobPhase,
    ACPGraduationStatus,
    ACPOnlineStatus
)
from virtuals_acp.contract_clients.contract_client_v2 import ACPContractClientV2
import pathlib
import os

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger("BuyerAgent")

env_path = pathlib.Path(__file__).parent.parent / ".env.local"
load_dotenv(dotenv_path=env_path, override=True)

class EnvSettings:
    BUYER_WALLET_PRIVATE_KEY = os.getenv("BUYER_WALLET_PRIVATE_KEY")
    BUYER_AGENT_WALLET_ADDRESS = os.getenv("BUYER_AGENT_WALLET_ADDRESS")
    BUYER_ENTITY_ID = int(os.getenv("BUYER_ENTITY_ID"))

def buyer():
    env = EnvSettings()

    def on_new_task(job: ACPJob, memo_to_sign: Optional[ACPMemo] = None):
        if (
            job.phase == ACPJobPhase.NEGOTIATION
            and memo_to_sign is not None
            and memo_to_sign.next_phase == ACPJobPhase.TRANSACTION
        ):
            logger.info(f"Paying for job {job.id}")
            job.pay_and_accept_requirement()
            logger.info(f"Job {job.id} paid")

            

        elif (
            job.phase == ACPJobPhase.TRANSACTION
            and memo_to_sign is not None
            and memo_to_sign.next_phase == ACPJobPhase.REJECTED
        ):
            logger.info(f"Signing job {job.id} rejection memo, rejection reason: {memo_to_sign.content}")
            memo_to_sign.sign(True, "Accepts job rejection")
            logger.info(f"Job {job.id} rejection memo signed")

        elif job.phase == ACPJobPhase.COMPLETED:
            logger.info(f"Job {job.id} completed, received deliverable: {job.deliverable}")

        elif job.phase == ACPJobPhase.REJECTED:
            logger.info(f"Job {job.id} rejected by seller")

    acp_client = VirtualsACP(
        acp_contract_clients=ACPContractClientV2(
            wallet_private_key=env.BUYER_WALLET_PRIVATE_KEY,
            agent_wallet_address=env.BUYER_AGENT_WALLET_ADDRESS,
            entity_id=env.BUYER_ENTITY_ID,
        ),
        on_new_task=on_new_task
    )

    # Browse available agents based on a keyword
    relevant_agents = acp_client.browse_agents(
        keyword="Chromia's EVAL",
        sort_by=[ACPAgentSort.SUCCESSFUL_JOB_COUNT],
        top_k=5,
        graduation_status=ACPGraduationStatus.ALL,
        online_status=ACPOnlineStatus.ALL,
    )
    logger.info(f"Relevant agents: {relevant_agents}")

    # Pick one of the agents based on your criteria (in this example we just pick the first one)
    chosen_agent = relevant_agents[0]
    # Pick one of the service offerings based on your criteria (in this example we just pick the first one)
    chosen_job_offering = chosen_agent.job_offerings[0]

    print("Sending Job")

    job_id = chosen_job_offering.initiate_job(
        service_requirement={ "details": "testing transaction" },
        expired_at=datetime.now() + timedelta(days=1), # expiredAt
    )
    logger.info(f"Job {job_id} initiated")
    logger.info("Listening for next steps...")

    threading.Event().wait()


if __name__ == "__main__":
    buyer()